credentials/insecure/insecure.go

```diff
@@ -0,0 +1,74 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package insecure provides an implementation of the
+// credentials.TransportCredentials interface which disables transport security.
+//
+// Experimental
+//
+// Notice: This package is EXPERIMENTAL and may be changed or removed in a
+// later release.
+package insecure
+
+import (
+	"context"
+	"net"
+
+	"google.golang.org/grpc/credentials"
+)
+
+// NewCredentials returns a credentials which disables transport security.
+func NewCredentials() credentials.TransportCredentials {
+	return insecureTC{}
+}
+
+// insecureTC implements the insecure transport credentials. The handshake
+// methods simply return the passed in net.Conn and set the security level to
+// NoSecurity.
+type insecureTC struct{}
+
+func (insecureTC) ClientHandshake(ctx context.Context, _ string, conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return conn, Info{credentials.CommonAuthInfo{SecurityLevel: credentials.NoSecurity}}, nil
```

Comment:
How does this impact call creds?  Will call creds not aware of security level be allowed?  Should they be?  What do the other languages do?  @yihuazhang do you know offhand?

Reply:
The security level of a channel is orthogonal to the security level of a call credential. A security level is an inherent property of both channel and call credential. For a call credential, its security level dictates the minimum security level a channel's security level should be in order to transfer the call credential. So in the case of insecure credentials, it can only transfer the call credentials having the security level - `NoSecurity`.  The above behavior is consistent across all languages (C-core, Java, and Go). PLMK if you have more questions.

Reply:
We have a bunch of call credentials that were written before we had the "security levels" concept.

Are those supposed to work with insecure creds?  I would assume anything that doesn't support security levels, because it is too old, should NOT be used with insecure creds.  At least, if they have `RequireTransportSecurity() = true`.

Did we add a check somewhere:

```go
if perRPCCreds.RequireTransportSecurity() && transportCreds.SecurityLevel() == INSECURE {
  // FAIL
}
```

I don't think we did.  Here are the places I can see where we check `RequireTransportSecurity`:

https://github.com/grpc/grpc-go/blob/9519efffeb5d1897ae8671568871a6d476986524/clientconn.go#L175-L182

and

https://github.com/grpc/grpc-go/blob/9519efffeb5d1897ae8671568871a6d476986524/internal/transport/http2_client.go#L559-L562
+
https://github.com/grpc/grpc-go/blob/9519efffeb5d1897ae8671568871a6d476986524/internal/transport/http2_client.go#L226-L237

In both cases, the *presence* of `TransportCredentials` is enough to make the `PerRPCCredentials` happy.  But that seems too weak now that we will have credentials that aren't actually providing security at all.  Or does "insecure" credentials mean "also fake out call credentials so they think the connection is secure"?  And if that *is* the case, these creds should return the max security level, shouldn't they?

Reply:
Good point. I agree the current check (if we leave as it is now) is too weak because as you said, all `PerRPCCredentials` having security level > `NoSecurity` can be communicated over channels created with insecure credentials. I think we should add the check you suggested above, and probably in `http2_client.go` where `GetRequestMetadata()` is called?

Reply:
I think we should also consider a channel's `Invalid` security level for the purpose of backward compatibility in the above check. Similar to `CheckSecurityLevel()` API in `credentials.go`

Reply:
Yes, that all sounds right to me.  But note that the check in `clientconn.go` can't be improved since the `TransportCredentials` don't report their security level until after the connection is established.  So the only place we can do this is in `http2_client.go`, and it will have to be a connection failure and not a channel-creation failure.

Also, do we require `!=NoSecurity` or `==PrivacyAndIntegrity` if `PerRPCCredentials.RequireTransportSecurity()==true`?  Let's move this conversation to #3974.